buildtools/rpm/rpm.go

```diff
@@ -0,0 +1,363 @@
+			IDMap[dep] = pkg.ID{
+				Type:     pkg.Raw,
+				Name:     locator.Project,
+				Revision: locator.Revision,
+			}
+			mapLock.Unlock()
+		}(d)
+	}
+	wg.Wait()
+
+	depGraph := graph.Deps{
+		Direct:     []pkg.Import{pkg.Import{Target: target, Resolved: IDMap[target]}},
+		Transitive: make(map[pkg.ID]pkg.Package),
+	}
+
+	for _, id := range IDMap {
+		imports := []pkg.Import{}
+
+		for _, dep := range depMap[id.Name] {
+			imports = append(imports, pkg.Import{
+				Target:   dep,
+				Resolved: IDMap[dep],
+			})
+		}
+
+		depGraph.Transitive[id] = pkg.Package{
+			ID:      id,
+			Imports: imports,
+		}
+	}
+
+	return depGraph, nil
+}
+
+// SystemPackages uploads license information for all system level dependencies.
+func (s Shell) SystemPackages() (graph.Deps, error) {
+	output, err := s.allPackagesRPM()
+	if err != nil {
+		return graph.Deps{}, err
+	}
+
+	lines := strings.Split(output, "\n")
+	dependencies := []dependency{}
+	for _, line := range lines {
+		dependencies = append(dependencies, dependencyFromOutput(line))
+	}
+
+	wg := sizedwaitgroup.New(runtime.GOMAXPROCS(0))
+	mapLock := sync.RWMutex{}
+	depGraph := graph.Deps{
+		Transitive: map[pkg.ID]pkg.Package{},
+	}
+	for _, d := range dependencies {
+		wg.Add()
+		go func(dep dependency) {
+			defer wg.Done()
+
+			locator, err := findAndUploadLicense(dep, s.LicenseDirectory, s.Upload)
+			if err != nil {
+				log.Warn(err.Error())
+			}
+
+			depID := pkg.ID{
+				Type:     pkg.Raw,
+				Name:     locator.Project,
+				Revision: locator.Revision,
+			}
+
+			mapLock.Lock()
+			depGraph.Transitive[depID] = pkg.Package{ID: depID}
+			depGraph.Direct = append(depGraph.Direct, pkg.Import{Target: locator.Project, Resolved: depID})
+			mapLock.Unlock()
+		}(d)
+	}
+	wg.Wait()
+
+	return depGraph, nil
+}
+
+func findAndUploadLicense(dep dependency, licenseDir string, upload bool) (fossa.Locator, *errors.Error) {
+	// 1. Try to upload the copyright files in the system license directory. 70% are usually here.
+	locator, err := fossa.UploadTarballDependency(filepath.Join(licenseDir, dep.name), upload, true)
+	if err != nil {
+		// 2. Upload the RPM declared license if the copyright files do not exist. This is normally always present.
+		var uploadErr error
+		locator, uploadErr = fossa.UploadTarballString(dep.name, dep.license, true, true, upload)
+		if uploadErr != nil {
+			return locator, &errors.Error{
+				Cause:           uploadErr,
+				Type:            errors.Unknown,
+				Troubleshooting: fmt.Sprintf("There was a problem uploading license information for %s. Ensure that you have a working network connection and that you can reach app.fossa.com.", dep.name),
+			}
+		}
+		log.Warnf("license files for `%s` could not be found in `%s`. The declared license `%s` will be used: %s", dep.name, licenseDir, dep.license, err.Error())
+	}
+
+	return locator, nil
+}
+
+// Parse dependency from output format dep-name,version,license.
+func dependencyFromOutput(line string) dependency {
```

Comment:
I'm not actually aware of a case where a dependency won't have a field filled, but I made sure to handle the event it does not. In the event it doesn't have one of these it will result in us logging a warning when we aren't able to find any dependencies.